[계산기] 이상현 미션 제출합니다.

idealHyun opened the pull request:
# 계산기 - 단위테스트

## 학습 내용
- Junit 사용법
- AssertJ 사용법

## 코드 설명
- 연산자는 Operator 라는 추상 클래스를 사용해서 구현
   - Operator를 상속해서 사칙연산인 4개의 클래스 구현
- 피연산자는 Operand 라는 클래스로 유효성 검증
- 문자열 계산은 Operator에서 2개 인자로 구현되게 정의한 바람에 재귀함수로 구현

## 회고
- 단계별로 브랜치를 파고 PR을 날리는 형식으로 했어야했는데 잊어버렸다.
- 단계별로 요구사항을 정리하면서 하려했는데 하다보니 잊어버렸다.
- 추상화에 대해 욕심을 내다가 중간에 코드의 흐름을 놓쳐버려서 코드가 많이 망가진 것 같다.
- 코드를 작성하면서 같이 테스트 작성을 하는 것이 아직 많이 미숙한 것 같다.
- 메인 메소드 없으니까 디버그 하는게 너무 힘들었는데 테스트 코드로 확인해야하는데 이것이 안되었다.
- 커스텀 구분자를 추가하기 위해서 마지막에 덧셈 Operator만 싱글톤으로 구현했는데 억지로 끼워맞춘것 같아서 아쉽다.
- 커스텀 구분자의 여부를 확인하는 로직이 `//` 로 시작하는 것을 체크했는데 다른 좋은 방법도 있을 것 같다.

## 궁금한 점
- 미션을 진행할 때 다음 단계 내용을 보고 구현하는게 좋을까요? 보더라도 모른다고 가정하고 구현하는게 좋을까요?

idealHyun committed:
docs: 1단계 README 작성

- 1단계 요구사항 정리

idealHyun committed:
feat: Calculator 클래스 생성

- 사칙연산 기능 구현
- 0으로 나눌 때 에러 메세지 띄우기

idealHyun committed:
test: Calculator 클래스 테스트 생성

- 사칙연산에 대한 테스트 생성

idealHyun committed:
docs: 2단계 요구사항 정리

idealHyun committed:
feat: Operand 클래스 생성

- 숫자를 모두 정수형으로 변환

idealHyun committed:
feat: Operator 인터페이스 및 사칙연산 클래스 구현

- Operator 인터페이스를 통해 사칙연산 구현

idealHyun committed:
feat: Operator 인터페이스 수정

- 연산자 get 함수 추가
- symbol을 여러개 가질 수 있게 배열로 구현

idealHyun committed:
feat: Operator 추상클래스로 변경

- symbol이 있는지 체크하는 함수 추가
- symbol 추가할 수 있는 함수 추가

idealHyun committed:
test: 문자열 입력 계산 테스트

idealHyun committed:
feat: 2단계 계산기 완성

idealHyun committed:
feat: , : 덧셈 부호로 추가

idealHyun committed:
fix: Operator 구현 수정

- List.of()로 하면 불변객체가 되어 추가를 못함

idealHyun committed:
fix: Operand 수정

- 음수나 숫자 이외의 값이 들어올 경우 RuntimeException 발생

idealHyun committed:
feat: 3단계 계산기 구현

- AddOperator 커스텀 구분자 추가로 인해 싱글턴으로 구현
- calculate 재귀함수로 구현

idealHyun committed:
feat: AssertJ 의존성 추가

idealHyun committed:
docs: 요구사항 정리

- 2단계 요구사항 변경
- 3단계 요구사항 정리

idealHyun committed:
feat: AssertJ 적용

idealHyun committed:
fix: AddOperator 수정

- 싱글톤 형식을 위해 public -> private

hong-sile commented on README.md:
```diff
@@ -1,3 +1,24 @@
 # java-calculator
 
-계산기 미션 저장소
+## 1단계 계산기 요구사항
```
요구사항을 정리하면서 작성한 것은 정말 잘해주셨습니다!

idealHyun replied:
다음 미션부터는 더 구체적으로 써보려고 노력하겠습니다!

hong-sile commented on src/test/java/domain/CalculatorTest.java:
```diff
@@ -0,0 +1,96 @@
+        new DivideOperator()
+    });
+
+    @Nested
```
Nested를 사용해주셨군요 상현님은 어떠한 목적으로 Nested를 사용하셨나요?

idealHyun replied:
큰 기능 아래의 하위 기능을 테스트하는 용도로 사용하려했습니다. 
1단계에서 인자 2개로 이루어지는 계산을 정수 계산과 문자열 계산으로 나누기 위해 Nested 를 사용하여 분리하려했었습니다.
정수 계산 테스트를 지우고 보니 Nested 때문에 쓸데 없이 한번 더 감싸진 것 같습니다.

hong-sile commented on src/main/java/domain/Operator.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class Operator {
```
Operator를 abstract class로 만드셨군요. 상현님은 abstract class를 어떨때 쓰시나요?

idealHyun replied:
인터페이스는 행위, 추상클래스는 공통적인 행위와 변수(상태?)를 가진 것을 제공하려고 할 때 사용하려고 하는 것 같습니다.
처음에는 Operator를 계산을 수행한다는 역할만 수행하면 될 것 같아서 인터페이스로 구현을 했다가 기호에 대한 변수값이 필요할 것 같아서 추상클래스로 변경했는데 이러한 이유로 변경을 했다면 맞는 흐름일까요?

hong-sile replied:
그런식으로 필요에 의해서 확장해나가는 방식은 좋다고 생각합니다. 👍 
초기에 인터페이스를 추가할때도 그런식으로 해보면 어떨까요?? 

초반엔 안 만들다가, 공통부분이 겹치고 바뀔것 같을 때 인터페이스를 도입해보는 방식으로요.

hong-sile commented on src/main/java/domain/Operand.java:
```diff
@@ -0,0 +1,25 @@
+
+    public Operand(String operand) {
+        try {
+            double parsed = Double.parseDouble(operand);
```
double로 parsing한 후에 int로 type casting을 했군요. 그러면 실제로 11.1 이런식으로 입력 되었을 떄 어떤 일이 일어날까요?

idealHyun replied:
소수점 아랫부분의 정보가 날아갑니다.
처음에 실수에 대해 계산을 처리하는 것을 목표로 했다가 `.` 도 구분자가 될수도 있다고 판단하고 정수들끼리만 계산이 가능하게 구현했습니다. 
결과 값이 정수로만 반환된다는 점에서 이렇게 결정했던 것 같습니다.

hong-sile commented on src/test/java/domain/CalculatorTest.java:
```diff
@@ -0,0 +1,96 @@
+        @Test
+        @DisplayName("문자열 입력 유효성 검증 테스트")
+        void validateInputExpressionTest() {
+            assertThatThrownBy(() -> calculator.calculateStringExpression("1+"))
```
테스트 메서드에 여러개의 assert가 존재하는 군요. 요런경우 어떤 문제가 발생할 수 있을까요?

idealHyun replied:
어떤 검증에서 실패했는지 구분이 어렵고, 중간에 테스트가 실패하는 경우 다음 케이스는 테스트가 이루어지지 않아 명확하게 알 수 없습니다.
다음 미션에서는 Test 어노테이션을 더 의미있게 사용해보겠습니다.

hong-sile replied:
👍 여러 개의 테스트를 SoftAssertions 으로 묶으면 하나의 테스트 메서드에서 할수도 있습니다.

하지만 이 케이스의 경우엔, ParameterizedTest를 활용해보는 것도 좋겠네요

hong-sile commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,87 @@
+            return 0;
+        }
+
+        int result;
```
요기서 int result를 별도로 선언하지 않고도, 처리하는 방법이 있을 것 같아요.

hong-sile commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,87 @@
+
+        int result;
+        if (expression.startsWith("//")) {
+            String remainedExpression = this.extractCustomSymbol(expression);
```
어떤 경우엔 this를 붙이고 메서드를 사용해주셨고, 어떤 경우엔 그냥 메서드를 호출해주셨군요. 

어떤 이유로 this를 붙이거나 붙이지 않으셨나요?

idealHyun replied:
이유없이 그냥 적었던것 같습니다

hong-sile commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,87 @@
+            if (matcher.group(1).length() != 1) {
+                throw new IllegalArgumentException("커스텀 구분자가 올바르지 않습니다.");
+            }
+            AddOperator.INSTANCE.addSymbol(matcher.group(1).charAt(0));
```
이 메서드 이름만보면 extract해서 반환만 해줄것 같은데 Operator에 addSymbol까지 해주고 있군요.

메서드 이름으론 이 동작을 추측할 순 없을 것 같아요.
이 동작을 별도로 분리하거나 메서드 이름을 바꿔보는 건 어떨까요?

idealHyun replied:
이 부분에 대해 수정해보겠습니다.

hong-sile commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,87 @@
+                    if (isInteger(rightExpression)) {
+                        rightOperand = new Operand(rightExpression);
+                    } else {
+                        int rightValue = calculate(rightExpression);
```
연산이 여러개 들어오는 걸 우측 operand를 재귀로 풀어내셨군요.
재밌는 방식인 것 같아요.

hong-sile commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.regex.Pattern;
+
+public class Calculator {
+    private static final String REGEX_PATTERN = "^\\/\\/(.)\\n([\\s\\S]*)";
```
regex는 강력하지만 그리 가독성이 좋은 방식은 아닙니다.
변수명에라도 이 패턴이 뭘 의미하는지 나타내주면 좀 더 읽기 쉬울 것 같아요.

idealHyun replied:
주석을 추가하거나 변수명에 최대한 의미를 담아보겠습니다.

hong-sile commented on src/test/java/domain/CalculatorTest.java:
```diff
@@ -0,0 +1,96 @@
+@DisplayName("계산기 테스트")
+public class CalculatorTest {
+
+    Calculator calculator = new Calculator(new Operator[] {
```
테스트에서 하나의 calculator와 operator들을 계속 쓰시는군요.

이런경우 어떤 문제가 발생할수 있을까요?

idealHyun replied:
calulator나 operator에서 가지는 상태값들을 제대로 다루지 못하면 다른 테스트들이 독립적으로(?) 실행되지 못해서 테스트 코드로서 역할을 다하지 못할 것 같습니다.

hong-sile replied:
맞아요! 테스트에서 상태가 유지되기 때문에 문제가 생길 수 있어요.

이 경우엔 custom opearotr를 등록하는 메서드에 대해서 테스트한 경우, 이게 다른 곳에도 영향을 줄 수 있을 수도 있습니다.
```
    @Test
    @DisplayName("커스텀 문자열 테스트")
    void calculateCustomSymbolTest() {
        assertAll(
            () -> assertThat(calculator.calculateStringExpression("")).isEqualTo(0),
            () -> assertThat(calculator.calculateStringExpression("//;\n1;4")).isEqualTo(5),
            () -> assertThat(calculator.calculateStringExpression("//;\n1;2;3")).isEqualTo(6)
        );
    }

```

요 메서드에서 저 상태값이 변경되니까요. 이런 케이스에 대한 인지를 하면서 짜면 훨씬 좋을 것 같아요..!

hong-sile reviewed (changes requested):
안녕하세요! 상현님 첫번쨰 미션 수고하셨습니다!

회고랑 요구사항까지 정리해서 작성해주신 것이 인상적이군요! 아주 좋습니다.

Operand랑 Operator를 만들어서 한게 되게 유니크하군요. 상현님이 클래스를 쪼개는 기준과 추상클래스를 사용하는 기준은 무엇인가요?
코드를 보면서 몇 가지 질문을 남겼어요. 확인해주고 답을 남겨주시면 좋을것 같아요.

일단 궁금한점에 대한 답변으론, 최대한 다음 단계 내용을 안 보고 모른다고 구현하는 것을 추천드리고 싶습니다.
미션을 수행해보셨으면 알겠지만, 미션은 큰 미션에서 단계별로 진행되는 형식입니다.
이 단계별로 할 때 기존의 것들을 마이그레이션하는 방향으로 진행되죠.

이 마이그레이션하는 것 조차 연습이 되고, 뒤에껄 모른 상태로 적당히 확장성 있는 구조로 짜는 것도 중요한 연습입니다.

그래서 저는 최대한 뒤에껄 모른다고 생각하고 미션을 수행하면 좋겠어요.

hong-sile commented on README.md:
```diff
@@ -1,3 +1,24 @@
+- 0으로 나누었을 경우 에러 메세지를 띄워준다.
+  - 두번째 인자가 0이라면 `0으로 나눌 수 없습니다.`를 띄워준다.
+
+## 2단계 계산기 요구사항
+
+- 문자열로 이루어진 수식이 들어오는 경우 계산이 이루어진다.
+  - 2개의 인자와 1개의 연산자가 들어온다.
+
+## 3단계 계산기
+
```
test

idealHyun committed:
fix: Test 코드 수정

- 같은 기능이여도 예외가 생기는 부분과 정상적인 결과를 뱉어내는 건 분리하기
- 여러개의 경우 assertAll()로 묶기
- Calculator @BeforeEach 로 Test 마다 독립성을 확보하기

idealHyun committed:
fix: Calculator 수정

- 정규표현식 추가 설명 및 변수 이름 변경
- 불필요한 변수(result) 제거
- 함수는 한 가지의 역할만 하기

idealHyun committed:
fix: Operand 수정

- 불필요한 코드 수정
